content/blog/semgrep-post-message/index.md

```diff
@@ -0,0 +1,175 @@
+      - pattern-not: |
+          function($OBJ){ if ($OBJ.origin !== $X) { ... } }
+      - pattern-not: |
+          function($OBJ){ if ($REGEX.test($OBJ.origin)) { ... } }
+```
+
+We want to match function declarations (`pattern`) that **do not** contain any of our safe patterns (`pattern-not`). A few things are worth noting here:
+
+1. Our metavariable is matching on the context captured by `$FUNC`
+2. We need the the `function($OBJ) { ... }` check to exclude external function declarations (our second scenario)
+3. The `$OBJ` metavariable is matching the function parameter and making sure the `.origin` is being called on it
+
+We could likely use a `metavariable` for our equality operators instead of declaring all of these `pattern-not`, but I like how explicit the rule currently is. Time to match those external functions!
+
+#### External functions
+
+```yaml
+pattern-either:
+  - pattern: |
+      function $FNAME(...) { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - pattern: |
+      $FNAME = (...) => { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+```
+
+_This is the only situation where I couldn't use the same pattern to match `arrow functions` and normal `function` declarations. Let me know if I could simplify this pattern._
+
+Here we are making sure that we can match any of the function declaration types with `pattern-either` and capturing the function names during declaration with `$FNAME`. Note that we use the same `$FNAME` metavariable to ensure that the **same name** is being used within `addEventListener`. The ellipsis (`...`) are there to match anything in-between those two lines.
+
+We are also capturing the context of the function body itself with `$CTX` and we will use it to exclude our **safe patterns**. Let's get to it:
+
+```yaml
+pattern-either:
+  - pattern: |
+      function $FNAME(...) { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - pattern: |
+      $FNAME = (...) => { $CTX }
+      ...
+      window.addEventListener('message', $FNAME,...)
+  - metavariable-pattern:
+    metavariable: $CTX
+    patterns:
+      - pattern-not: |
+          if ($OBJ.origin == $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin === $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin != $X) { ... }
+      - pattern-not: |
+          if ($OBJ.origin !== $X) { ... }
+      - pattern-not: |
+          if ($REGEX.test($OBJ.origin)) { ... }
+```
+<br />
+
+Nothing new here, we are ensuring that we ignore matches if they contain any of the `if` patterns declared by our `pattern-not` operators. These patterns are being checked on the context of `$CTX` which is our function body.
+
+With those two cases being taken care of separately we can move to our final rule declaration.
+
+## Final pattern
+
+In order to make this work we need to glue both scenario using `pattern-either`. Let's check our final rule in the playground:
```

Comment:
By the way, for things that are meaningfully different, you could just keep them as separate rules and then run them both together with `semgrep --config path/to/dir_with_rules` or `semgrep --config semgrep.dev/URL --config.dev/otherURL`.

It's a stylist thing, you can't put everything together, our thought process is mostly: are these meaningfully different issues that should be treated separately? Will it be easier to review and understand this rule if we catch the cases separately?

Just FYI, don't feel like you need to change anything.

Reply:
That's a great point! For example, we could have a rule like this one for origin checking when receiving messages and a separate one for the target origin when actually posting the message.